### ISSUE TITLE
Inconsistency in the base types accepted by the ** operator

### ISSUE THREAD

**Issue by magus237** [ethereum/solidity#8895]
Opening this issue after chatting with @chriseth.

Consider the following function:
```
  function f() public payable {
      int x_int = -2;
      uint x_uint = 2;
      uint y_uint = 4;
      int z1 = x_int ** y_uint; // Typecheck OK (with Solidity >=0.6.0, but NOK otherwise)
      int z2 = (-2) ** y_uint; // Typecheck NOK
      uint z3 = x_uint ** y_uint; // Typecheck OK
      uint z4 = (2) ** y_uint; // Typecheck OK
  }
```

I'm surprised that z2 is rejected while z1 and z4 are accepted, but according to @chriseth, both z2 and z4 should be rejected anyways, because (quoting) "`**` needs the base type to be explicit".

**Comment by chriseth** [comment]
It turns out that if  `2<<x` and `2**x` are valid, the type of the resulting expression is always the type of `x`. This is certainly wrong, because for `x<<y` and `x**y`, the type of the expression is always the type of `x`.

The reason behind this bug is clear, the only question is what the desired behaviour should be.

The bug is also not too bad, because we have warnings in place if the result of the expression is not a 256 bit type.

**Comment by axic** [comment]
I think my intuition is when the lhs is a literal it suggests we expand to a type which "fits the expression", e.g. `uint256`. Unless the lhs is explicitly typed.

**Comment by chriseth** [comment]
Yeah, maybe we just do that and rely on the warning.

**Comment by chriseth** [comment]
Fixed by #8994
